### DISCUSSION TITLE
Range of arguments lengths / nargs in python click

### DISCUSSION THREAD

**Discussion by sylvorg** [pallets/click#2747] [Q&A]
Hello!

Is there a way to allow a range of arguments in click using `nargs`, such exactly between 2 and 4 arguments? Something like the following, for example, in a `test.py` file:

```python
@click.group()
@click.argument("ints", nargs=(2, 4))
def test(ints):
    ...

test()
```

Which would then allow:

```bash
./test.py 1 2
./test.py 1 2 3
./test.py 1 2 3 4
```

But disallow:

```bash
./test.py 1
./test.py 1 2 3 4 5
```

I'm expecting this to work similar to `nargs=-1`, where it swallows everything after it.

Thank you kindly for the help!

**Comment by davidism** [comment]
No, because it leads to ambiguities in parsing. Let's say we parse 4 tokens then find we need one more for the next arg. We'd need to implement backtracking to the reset the parse state, consume 3, then try again. Backtracking would add a ton of complexity and be slower. It would get worse with multiple ranged arguments or optional arguments. And impossible with subcommands. That's why you can only have one greedy argument, and optional arguments always fill left to right.

**Reply by sylvorg** [reply]
I wasn't quite intending it to be used _with_ `nargs=-1`, but be it's own thing, so you'd only be and to use one or the other. Basically, we count how many extra arguments have already been processed, and if it doesn't fall in the specified range, raise an error.

Would using a callback with `nargs=-1` be more appropriate?
